tremor-script/src/ast/analyzer.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020-2021, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{
+    ArrayPattern, ArrayPredicatePattern, AssignPattern, Expression, ImutExprInt, Pattern,
+    PredicateClause, PredicatePattern, RecordPattern, TestExpr, TuplePattern,
+};
+
+pub(crate) trait Costly {
+    fn cost(&self) -> u64;
+}
+
+impl Costly for TestExpr {
+    fn cost(&self) -> u64 {
+        self.extractor.cost()
+    }
+}
+
+impl<'script, Ex: Expression + 'script> Costly for PredicateClause<'script, Ex> {
+    fn cost(&self) -> u64 {
+        let g = if self.guard.is_some() { 10 } else { 0 };
+        g + self.pattern.cost()
+    }
+}
+
+impl<'script> Costly for Pattern<'script> {
+    fn cost(&self) -> u64 {
+        match self {
+            Pattern::DoNotCare | Pattern::Default => 0,
+            Pattern::Expr(_) => 10,
```

Comment:
could you extract those costs to some constants with meaningful names, so the reason for the cost estimation becomes clearer? That would help me and other noobs understand. :)

Reply:
not sure how much it'll help but I'll add some, at least it'll make adjusting them easier later :D :+1: